### PR TITLE
Merge Preclusion and Prerequisite view into Integration branch

### PR DESCRIPTION
**Pull request by nlzz22**
This branch implemented User Story Queries number 2:
"See all modules associated with a module"

Branch: queries2
Implemented: You can now see the prereqs and preclusions of a module at the module view page.

Note that: Editing prereqs and preclusions is not available yet and they are reflected as not available at the edit general module info page.

### CODE REVIEW

**Comment by tgqiang** [comment]
@helloqx You did the splitting of module info into their respective chunks of info with your jupyter notebook right? Something seems truncated here, was it originally already truncated? If not, you might want to re-look your jupyter script again.

image

**Comment by tgqiang** [comment]
> Branch: queries2
> Implemented: You can now see the prereqs and preclusions of a module at the **individual module view page**.

@nlzz22 You mean **view module page**? I don't find them in the page you specified.

**Comment by tgqiang** [comment]
@nlzz22 @helloqx @xaterz I'm not sure if you all want to, additionally, store the prose form of the prerequisites/preclusions, then render additionally the prose form of these info in the 'View Module' page, seeing that the ones in the database have the non-module-related prerequisites/preclusions portions **discarded**.

Current display of module preclusions and prerequisites are pretty much fine for now.

File: components/model.py [hunk 1]
```diff
@@ -926,6 +925,104 @@ def get_prerequisite(module_code):
...
+
+    # sort list of lists based on index (which is the first elem of each row)
+    prereq_list.sort(key=lambda row: row[INDEX_FIRST_ELEM])
+
+    prereq_string = convert_list_of_prereqs_to_string(prereq_list)
+
+    return prereq_string
+
+
+def convert_list_of_prereqs_to_string(prereq_list):
```
Comment (tgqiang) [inline components/model.py]: Your algorithm is wrong.

Sample fail case:
**For module CP3880**

Expected:
[ IS2102 or CG1413 and (CS2103 or IS2103 or IS2150)] and preferably completed [IS3102 or (CS3201 and CS3202) or (CS3281 and CS3282) or IS4102 or (CS4201 and CS4202) or (CS4203 and CS4204) or (CG2001). Student Selection process will be enforced.
(my parser ignored all clauses after 'preferably completed ...' as that clause does not form the true prerequisites for this module.)

Actual:
image

Diff:

**Expected**: IS2102 or CG1413 ***and (CS2103 or IS2103 or IS2150)***
**Actual**: IS2102 or CG1413 ***or CS2103 or IS2103 or IS2150***

**Comment by nlzz22** [comment]
> @nlzz22 You mean view module page? I don't find them in the page you specified.
Yes, updated the pull request to reflect "view module" page instead

File: components/model.py (hunk 1 above)
Comment (nlzz22) [inline components/model.py]: There is something wrong with my prerequisitePopulator.sql, will push a fix soon

File: components/handlers/module_view_in_ay_sem.py [hunk 2]
```diff
@@ -119,12 +119,14 @@ def GET(self):
...
         overlapping_mod_list = model.get_mod_taken_together_with(module_code)
+        prereq_string = model.get_prerequisite_as_string(module_code)
+        preclude_string = model.get_preclusion_as_string(module_code)
         is_starred = model.is_module_starred(module_code, web.cookies().get('user'))
 
         return RENDER.individualModuleInfo(module_info, is_future_ay,
                                            target_ay_sem, self.mounting_status,
                                            self.quota, overlapping_mod_list,
-                                           is_starred)
+                                           is_starred, prereq_string, preclude_string)
```
Comment (xaterz) [inline components/handlers/module_view_in_ay_sem.py]: Does module view in ay sem need prereq and preclusion data? Currently the data are passed to the UI but the UI does not render or use the data in any way.

Comment (xaterz) [inline components/handlers/module_view_in_ay_sem.py]: Does module view in ay sem need prereq and preclusion data? Currently the data are passed to the UI but the UI does not render or use the data in any way.

File: templates/individualModuleInfo.html [hunk 3]
```diff
@@ -1,4 +1,4 @@
-$def with (moduleInfo, isFutureAY, aySem, mountingStatus, quota, overlappingList, isStarred)
+$def with (moduleInfo, isFutureAY, aySem, mountingStatus, quota, overlappingList, isStarred, prereq, preclude)
```
Comment (xaterz) [inline templates/individualModuleInfo.html]: Again, does this page need prereq and preclusion?

**Comment by nlzz22** [comment]
Removed unused instances of prereq and preclusion string as per comments. @xaterz  Pls review again, thanks

**Pull request merged**